contextual.go

```diff
@@ -65,22 +65,54 @@ var (
 // To remove a backing logr implemention, use ClearLogger. Setting an
 // empty logger with SetLogger(logr.Logger{}) does not work.
 //
+// If the logger instance implements the ContextualLogger interface in addition
+// to logr.Logger, then the instance will also be called directly
+// by code that retrieves a logger with FromContext, TODO, or Background.
+// In that case the logger instance should do its own verbosity checks
+// in Enabled because klog will not do that for it.
+//
+// AddSupportsContext can be used for loggers which don't know about this
+// klog specific convention.
+//
 // Modifying the logger is not thread-safe and should be done while no other
 // goroutines invoke log calls, usually during program initialization.
 func SetLogger(logger logr.Logger) {
 	globalLogger = &logger
 	contextualLogger = false
+	if contextual, ok := logger.GetSink().(ContextualLogSink); ok && contextual.SupportsContext() {
+		contextualLogger = true
+	}
 }
 
-// SetContextualLogger does the same as SetLogger, but in addition the
-// logger may also get called directly by code that retrieves it
-// with FromContext, TODO or Background. The logger therefore must
-// implements its own verbosity checking.
-func SetContextualLogger(logger logr.Logger) {
-	globalLogger = &logger
-	contextualLogger = true
+// ContextualLogSink is an extension of the logr.LogSink interface that
+// indicates to SetLogger whether the logger may get called directly by
+// application code. Only the LogSink seen by SetLogger needs to implement
+// this, clones of it as they occur during logger.V, logger.WithName or
+// logger.WithValues don't need it.
+type ContextualLogSink interface {
+	logr.LogSink
+
+	// SupportsContext returns true if the logger may get called directly
+	// by application code.
+	SupportsContext() bool
+}
+
+// AddSupportsContext wraps a logger that doesn't know about ContextualLogSink
+// so that SetLogger knows that it can also be called directly.
+func AddSupportsContext(logger logr.Logger) logr.Logger {
+	return logr.New(contextualLogSinkWrapper{logger.GetSink()})
```

Comment:
This is slightly broken: we loose the [`level` field](https://github.com/go-logr/logr/blob/28755ae811e8511239e62a7b90b5ca808b691fca/logr.go#L241).

If someone does `SetLogger(AddSupportedContext(logger.V(1)))`, then what we instead is `SetLogger(AddSupportedContext(logger))`.

I can see a way around that, but it's going to be very ugly.

Reply:
No, my idea doesn't work. I wanted to call logger.Enabled and then see how LogSink.Enabled is called (it gets a level), but it's not possible to intercept that call.